lms/djangoapps/instructor_task/tasks_helper/grades.py

```diff
@@ -307,8 +314,40 @@ def __init__(self, _xmodule_instance_args, _entry_id, course_id, _task_input, ac
         self.course_id = course_id
         self.report_for_verified_only = generate_grade_report_for_verified_only()
         self.task_progress = TaskProgress(self.action_name, total=None, start_time=time())
-        self.course = get_course_by_id(self.course_id)
         self.file_name = 'problem_grade_report'
+        self.log_additional_info = six.text_type(self.course_id) == 'course-v1:MITx+CTL.SC0x+1T2019'
+
+    @lazy
+    def course(self):
+        return get_course_by_id(self.course_id)
+
+    @lazy
+    def graded_scorable_blocks_header(self):
+        """
+        Returns an OrderedDict that maps a scorable block's id to its
+        headers in the final report.
+        """
+        scorable_blocks_map = OrderedDict()
+        grading_context = grades_context.grading_context_for_course(self.course)
+        for assignment_type_name, subsection_infos in six.iteritems(grading_context['all_graded_subsections_by_type']):
+            for subsection_index, subsection_info in enumerate(subsection_infos, start=1):
+                for scorable_block in subsection_info['scored_descendants']:
+                    header_name = (
+                        "{assignment_type} {subsection_index}: "
+                        "{subsection_name} - {scorable_block_name}"
+                    ).format(
+                        scorable_block_name=scorable_block.display_name,
+                        assignment_type=assignment_type_name,
+                        subsection_index=subsection_index,
+                        subsection_name=subsection_info['subsection_block'].display_name,
+                    )
+                    scorable_blocks_map[scorable_block.location] = [header_name + " (Earned)",
+                                                                    header_name + " (Possible)"]
+        return scorable_blocks_map
+
+    @lazy
+    def course_structure(self):
+        return get_course_in_cache(self.course_id)
```

Comment:
Why this is needed? This will be getting all the course structure again.

Reply:
I can't understand by `again`? what is the other place we are loading the structure?

Reply:
It returns the block structure in the `cache` for the course. `self.course` will return the course from modulestore.

Reply:
ok.